### ISSUE TITLE
Fields that resolve with post meta not showing in Previews

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#1615]
----
See: https://github.com/gatsbyjs/gatsby-source-wordpress-experimental/issues/324
----

It's been reported that when registering a field using `register_graphql_field` and resolving the field using `get_post_meta()` the field isn't properly returning meta when querying a post as a Preview.

**Comment by jasonbahl** [comment]
@stupearce @TylerBarnes from what I can tell, things are working as expected in WPGraphQL when it comes to adding fields that resolve as post_meta. 

To replicate, we need to do 2 things: 

- add a meta box to input data on Posts
- add a field to the GraphQL Schema that resolves with the meta

## Add a Meta Box
With the following code I add a meta box to the `post` Post Type:

```php

// Add the metabox
add_action( 'add_meta_boxes', 'wpdocs_register_meta_boxes' );
function wpdocs_register_meta_boxes() {
	add_meta_box( 'wpgraphql-test', __( 'WPGraphQL Test', 'textdomain' ), 'render_wp_graphql_test_metabox', 'post' );
}

// Render the metabox
function render_wp_graphql_test_metabox( $post ) {
	?>
	<label for="wp_graphql_test">Test Field...</label>
	<input type="text" name="wp_graphql_test" value="<?php echo get_post_meta( $post->ID, 'wp_graphql_test', true ); ?>" >
	<?php
}

// Save the data (in production, you'd verify nonces, etc too)
add_action( 'save_post', function( $post_id ) {
	if ( isset( $_POST['wp_graphql_test'] ) ) {
		update_post_meta( $post_id, 'wp_graphql_test', $_POST['wp_graphql_test'] );
	}
});
```

This gives me a metabox with a text field on my post edit screen like so: 

![Screen Shot 2020-12-08 at 10 51 02 AM](https://user-images.githubusercontent.com/1260765/101521574-451c3280-3943-11eb-93ab-f4857d029d03.png)

I can publish a post with this field populated:

![Screen Shot 2020-12-08 at 10 52 14 AM](https://user-images.githubusercontent.com/1260765/101521718-709f1d00-3943-11eb-8bac-acc724153e35.png)

And then check the database to see the data was saved as meta for the post:

![Screen Shot 2020-12-08 at 10 53 09 AM](https://user-images.githubusercontent.com/1260765/101521829-90364580-3943-11eb-9a41-65fb7ca268da.png)

## Add the field to GraphQL

I can then add the field to the WPGraphQL Schema like so: 

```php
add_action( 'graphql_register_types', function() {

	register_graphql_field( 'Post', 'my_test_field', [
		'type' => 'String',
		'description' => __( 'My test field', 'wp-graphql' ),
		'resolve' => function( $post ) {
			return get_post_meta( $post->databaseId, 'wp_graphql_test', true );
		}
	]);

} );
```

And then I can query like so: 

```graphql
{
  post(id: 1938, idType: DATABASE_ID) {
    id
    title
    my_test_field
  }
}
```

And get the following results: 

```json
{
  "data": {
    "post": {
      "id": "cG9zdDoxOTM4",
      "title": "Test meta",
      "my_test_field": "test meta..."
    }
  }
}
```

Screenshot of executed query in GraphiQL:

![Screen Shot 2020-12-08 at 10 55 14 AM](https://user-images.githubusercontent.com/1260765/101522067-f6bb6380-3943-11eb-925d-d1900fc83469.png)

## Query post as a preview

I can then make changes to the post:

![Screen Shot 2020-12-08 at 12 42 22 PM](https://user-images.githubusercontent.com/1260765/101533392-01c9c000-3953-11eb-9b91-99d103a4e142.png)

Then click preview, then query with GraphQL again. This time, I have a query for the published post and the preview of the post: 

```graphql
{
  published: post(id: 1938, idType: DATABASE_ID, asPreview: false) {
    id
    title
    content
    my_test_field
  }
  preview: post(id: 1938, idType: DATABASE_ID, asPreview: true) {
    id
    title
    content
    my_test_field
  }
}
```

And in response I get the following:

```json
{
  "data": {
    "published": {
      "id": "cG9zdDoxOTM4",
      "title": "Test meta",
      "content": null,
      "my_test_field": "test meta..."
    },
    "preview": {
      "id": "cG9zdDoxOTQx",
      "title": "Test meta",
      "content": "<p>Changes to the post</p>\n",
      "my_test_field": "test meta..."
    }
  }
}
```

![Screen Shot 2020-12-08 at 12 42 36 PM](https://user-images.githubusercontent.com/1260765/101533361-f70f2b00-3952-11eb-9b83-edb44b15845e.png)

Here we can see that the published content is still null, but the preview content includes the changes I made before clicking preview.

And in both cases, the meta is being returned. 

WordPress doesn't version post meta, so WPGraphQL is resolving the meta for the revision using the parent (the published post). 

If you have a mechanism for storing revised meta (some plugins like ACF do this) you can use the `graphql_resolve_revision_meta_from_parent` filter, as seen [here](https://github.com/wp-graphql/wp-graphql/blob/29cae903dd40739d5bac80d9100ad921089f0127/src/Model/Post.php#L268) and [here](https://github.com/wp-graphql/wp-graphql/blob/d72ca3088a47bf51d767a040311041f5e581d60b/tests/wpunit/PreviewTest.php#L275-L386).

**Comment by jasonbahl** [comment]
Closing in favor of: https://github.com/gatsbyjs/gatsby-source-wordpress-experimental/issues/324#issuecomment-741050789

**Comment by jasonbahl** [comment]
@stupearce welp, upon further investigation, @TylerBarnes and I discovered that there IS indeed a bug in WPGraphQL. 

When multiple fields that resolve from post meta are asked for on a preview, the first field resolves correctly, but the others do not. 

In my reproduction above had one field and it was working. 

If I take the same reproduction, but query other fields that resolve from meta first, this field no longer resolves properly. 

Here's an example of the query working as I'd expect it to. We see the `propertyRef` resolving as expected for the published node and the preview node.

![Screen Shot 2020-12-09 at 7 41 47 AM](https://user-images.githubusercontent.com/1260765/101644068-01343680-39f2-11eb-9439-5d0b760b20b4.png)

However, if I query the `enclosure` field before the `propertyRef` field, then the `propertyRef` field returns a "0". 

![Screen Shot 2020-12-09 at 7 43 40 AM](https://user-images.githubusercontent.com/1260765/101644308-42c4e180-39f2-11eb-9932-d3b80d2ccb21.png)

This indeed appears to be a bug with how meta is resolved for previews.

**Comment by stupearce** [comment]
@jasonbahl @TylerBarnes thanks so much for persevering with this once, much appreciated hopefully now you got a repro test it should be plain sailing :-)

**Comment by jasonbahl** [comment]
@stupearce can you update to WPGraphQL v1.0.1. I believe this should be resolved. I have test cases for it and was able to reproduce the issue and solution locally, but would love to also be able to confirm on your staging site.

**Comment by jasonbahl** [comment]
@stupearce I believe this issue is resolved with WPGraphQL v1.0.1. 

I'm going to close this issue. If you update and are still experiencing issues with this, please open a new Github issue with any added context that might be helpful to reproduce, and reference this issue for additional context.
